### PR TITLE
Docs for `GraphBuilder` component

### PR DESCRIPTION
**Pull request by rutujasurve94**
Add usage docs for:
- [x] `GraphBuilder`

### CODE REVIEW

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 1]
```diff
@@ -9,6 +9,69 @@ import com.typesafe.config.Config
...
+  * {s}`GraphBuilder`
+  *
+  * : `GraphBuilder` trait for creating a Graph by adding and deleting vertices and edges.
+  *    While using `GraphBuilder`, override {s}`parseTuple(tuple: T)` to define parsing of rows and use methods {s}`addVertex`, {s}`addEdge`
+  *    for adding vertices and edge to the graph. Deletion can be performed with {s}`deleteEdge` and {s}`deleteVertex` methods.
+  *
+  * ## Methods
+  *
+  *    {s}`addVertex(updateTime: Long, srcId: Long)`
+  *      : Adds vertex for a timestamp {s}`updateTime` and ID {s}`srcId`
```
Comment (Haaroon) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:22]: yes this adds a vertex, but it does more, for example it does not incvlude properties and it does not include the type? but what does add vertex actually do? it adds  the object `addVertex` to the updates.

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 2]
```diff
@@ -9,6 +9,69 @@ import com.typesafe.config.Config
...
+  *
+  *    {s}`addVertex(updateTime: Long, srcId: Long, properties: Properties)`
+  *      : Adds vertex for a timestamp {s}`updateTime` and ID {s}`srcId`, with Properties {s}`properties`
```
Comment (Haaroon) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:25]: you can probobaly just add a long description to one of the addvertex functions, the one likely with the most arguments

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 3]
```diff
@@ -9,6 +9,69 @@ import com.typesafe.config.Config
...
+  *
+  *    {s}`addVertex(updateTime: Long, srcId: Long, vertexType: Type))`
+  *      : Adds vertex for a timestamp {s}`updateTime` and ID {s}`srcId` for vertex type {s}`vertexType`
+  *
+  *    {s}`addVertex(updateTime: Long, srcId: Long, properties: Properties, vertexType: Type)`
+  *      : Adds vertex for timestamp {s}`updateTime`, ID {s}`srcId`, Properties {s}`properties`, vertex type {s}`vertexType`
+  *
+  *    {s}`deleteVertex(updateTime: Long, srcId: Long)`
+  *      : Deletes vertex for timestamp {s}`updateTime`, ID {s}`srcId`
```
Comment (Haaroon) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:34]: what does a delete vertex do? its doesnt delete the actual vertex right? it creates a new " vertex delete" object  that a specific time, removes the vertex from the graph scope, the vertex is still available in previous scopes, this object is sent as an update?

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 4]
```diff
@@ -9,6 +9,69 @@ import com.typesafe.config.Config
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
+/**
+  * {s}`GraphBuilder`
+  *
+  * : `GraphBuilder` trait for creating a Graph by adding and deleting vertices and edges.
+  *    While using `GraphBuilder`, override {s}`parseTuple(tuple: T)` to define parsing of rows and use methods {s}`addVertex`, {s}`addEdge`
```
Comment (Haaroon) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:16]: this also creates a bunch of updates, we should also explain these. what does a  new update represent, where is it stored, when are they used, who retrieves and processes them? 

missing the docs for the getUpdates function, what does that do ?

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 5]
```diff
@@ -9,6 +9,69 @@ import com.typesafe.config.Config
...
+  *      : Adds edge between ID {s}`srcId` and {s}`dstId` for timestamp {s}`updateTime`
+  *
+  *    {s}`addEdge(updateTime: Long, srcId: Long, dstId: Long, properties: Properties)`
+  *      : Adds an edge between ID {s}`srcId` and {s}`dstId` for timestamp {s}`updateTime`, Properties {s}`properties`
+  *
+  *    {s}`addEdge(updateTime: Long, srcId: Long, dstId: Long, edgeType: Type)`
+  *      : Adds an edge between ID {s}`srcId` and {s}`dstId` for timestamp {s}`updateTime` with Edge Type {s}`edgeType`
+  *
+  *    {s}`deleteEdge(updateTime: Long, srcId: Long, dstId: Long)`
+  *      : Deletes an edge between ID {s}`srcId` and {s}`dstId` for timestamp {s}`updateTime`
```
Comment (Haaroon) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:46]: same comments applies to this from the previous i made above.

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 6]
```diff
@@ -9,6 +9,69 @@ import com.typesafe.config.Config
...
+  *
+  *    {s}`addEdge(updateTime: Long, srcId: Long, dstId: Long, properties: Properties, edgeType: Type)`
+  *      : Adds an edge between ID {s}`srcId` and {s}`dstId` for timestamp {s}`updateTime` with Edge Type {s}`edgeType`, Properties {s}`properties`
+  *
+  *    {s}`parseTuple(tuple: T): Unit`
+  *      : Parse row {s}`tuple` to extract source node, destination node, timestamp info, etc.
```
Comment (Haaroon) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:52]: the parse tuple is a very important function, what does it do? it is used to process the messages from the spout? where is a short example? what does it return?

**Review by Haaroon — CHANGES_REQUESTED** [review]
this documentation is far too simple, as core developers we understand what raphtory does, but what about new people? 

also what about the `BuilderExecutors.scala` and the `GraphMessages.scala` files, they are apart of the graph builder core.

**Comment by rutujasurve94** [comment]
Thanks a lot for the review @Haaroon , I've added detailed docs as per your review. I'll update soon with PRs on `BuilderExecutors.scala` and `GraphMessages.scala`. 👍

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 7]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
...
+/**
+  * {s}`GraphBuilder`
+  *
+  * : `GraphBuilder` trait for creating a Graph by adding and deleting vertices and edges.
+  *    While using `GraphBuilder`, override {s}`parseTuple(tuple: T)` to define parsing of rows and use methods {s}`addVertex`, {s}`addEdge`
+  *    for adding vertices and edge to the graph. Deletion can be performed with {s}`deleteEdge` and {s}`deleteVertex` methods.
+  *
+  *    `GraphBuilder` creates a list of updates {s}`GraphUpdate` representing objects signifying vertex/edge add/delete
+  *     along with graph alteration information like update timestamp
+  *     These updates are retrieved and processed by the `Writer` and stored into `GraphPartition` as `avroSchema`
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:21]: Schemas are not used any more!

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 8]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
...
+  *     at the time of ingesting vertices, edges and processing updates.
+  *
+  * ## Methods
+  *
+  *    {s}`getUpdates(tuple: T)(failOnError: Boolean = true): List[GraphUpdate]`
+  *      : Parses `tuple` and fetches list of updates for the graph {s}`GraphUpdate`
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:27]: This is used internally to retrieve the updates. You should make it clear that users are not normally expected to override this!

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 9]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
...
+  *
+  *    {s}`addVertex(updateTime: Long, srcId: Long, properties: Properties, vertexType: Type)`
+  *      : Adds object {s}`addVertex` to the list of graph updates, signifies vertex addition for timestamp {s}`updateTime`, ID {s}`srcId`, Properties {s}`properties`, vertex type {s}`vertexType`
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:30]: there is no object called `addVertex`. Explain how to use Properties and link to appropriate documentation.

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 10]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
+/**
+  * {s}`GraphBuilder`
+  *
+  * : `GraphBuilder` trait for creating a Graph by adding and deleting vertices and edges.
+  *    While using `GraphBuilder`, override {s}`parseTuple(tuple: T)` to define parsing of rows and use methods {s}`addVertex`, {s}`addEdge`
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:16]: Explain where the input for parseTuple is coming from and link to spout documentation

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 11]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
...
+  * ## Methods
+  *
+  *    {s}`getUpdates(tuple: T)(failOnError: Boolean = true): List[GraphUpdate]`
+  *      : Parses `tuple` and fetches list of updates for the graph {s}`GraphUpdate`
+  *
+  *    {s}`addVertex(updateTime: Long, srcId: Long, properties: Properties, vertexType: Type)`
+  *      : Adds object {s}`addVertex` to the list of graph updates, signifies vertex addition for timestamp {s}`updateTime`, ID {s}`srcId`, Properties {s}`properties`, vertex type {s}`vertexType`
+  *
+  *    {s}`deleteVertex(updateTime: Long, srcId: Long)`
+  *      : Creates a {s}`deleteVertex` object that at a specific time {s}`updateTime`, removes the vertex ID {s}`srcId` from the graph scope. The vertex is still available in previous scopes.
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:33]: no `deleteVertex` object either

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 12]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
...
+  *
+  *    {s}`addEdge(updateTime: Long, srcId: Long, dstId: Long, properties: Properties, edgeType: Type)`
+  *      : Adds object {s}`addEdge` to the list of graph updates, signifies edge addition for an edge between ID {s}`srcId` and {s}`dstId` for timestamp, {s}`updateTime` with Edge Type {s}`edgeType`, Properties {s}`properties`
+  *
+  *    {s}`assignID(uniqueChars: String): Long`
+  *      : Assigns long hash for string {s}`uniqueChars`
+  *
+  *    {s}`parseTuple(tuple: T): Unit`
+  *      : Processes messages {s}`tuple` representing a singular row from the spout to extract source node, destination node, timestamp info, etc.
+  *        Creates vertices and edges with the parsed information, find example usage below.
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:47]: This is the most important function, move it to the top of the list!

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 13]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
...
+  *
+  *    {s}`deleteVertex(updateTime: Long, srcId: Long)`
+  *      : Creates a {s}`deleteVertex` object that at a specific time {s}`updateTime`, removes the vertex ID {s}`srcId` from the graph scope. The vertex is still available in previous scopes.
+  *        This object is sent as an update
+  *
+  *    {s}`deleteEdge(updateTime: Long, srcId: Long, dstId: Long)`
+  *      : Creates a {s}`deleteEdge` object that at a specific time {s}`updateTime`, removes the edge between {s}`srcId` and {s}`dstId` from the graph scope. The edge is still available in previous scopes.
+  *
+  *    {s}`addEdge(updateTime: Long, srcId: Long, dstId: Long, properties: Properties, edgeType: Type)`
+  *      : Adds object {s}`addEdge` to the list of graph updates, signifies edge addition for an edge between ID {s}`srcId` and {s}`dstId` for timestamp, {s}`updateTime` with Edge Type {s}`edgeType`, Properties {s}`properties`
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:40]: same here

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala [hunk 14]
```diff
@@ -9,6 +9,66 @@ import com.typesafe.config.Config
...
+  *
+  *    {s}`assignID(uniqueChars: String): Long`
+  *      : Assigns long hash for string {s}`uniqueChars`
```
Comment (ljeub-pometry) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphBuilder.scala:43]: Explain that this is optional and only provided for convenience

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]

**Comment by rutujasurve94** [comment]
@ljeub-pometry I've updated the PR as per your review. Thanks a lot 👍

**Comment by ljeub-pometry** [comment]
I cleaned up the documentation and added documentation for the different Properties classes. However, this meant a minor refactor which changes the import for the properties (they live inside a new Properties object so you can do `import com.raphtory.core.components.graphbuilder.Properties._` to get them but this means any `GraphBuilder` implementation needs a small edit.

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphMessages.scala [hunk 15]
```diff
@@ -2,21 +2,72 @@ package com.raphtory.core.components.graphbuilder
...
-/** @DoNotDocument */
-sealed trait Property {
-  def key: String
-  def value: Any
+import Properties._
+
+/**
+  * # Properties
+  *
+  * Vertex/Edge properties
```
Comment (Haaroon) [inline src/main/scala/com/raphtory/core/components/graphbuilder/GraphMessages.scala:10]: please add a short line/para explaining what properties are for,

**Review by Haaroon — CHANGES_REQUESTED** [review]

**Comment by rutujasurve94** [comment]
Thanks a lot @ljeub-pometry.
`Properties` to all `GraphBuilder` implementations in a refactoring PR post this. Are we good to merge?

File: src/main/scala/com/raphtory/core/components/graphbuilder/GraphMessages.scala (hunk 15 above)
Comment (rutujasurve94) [reply src/main/scala/com/raphtory/core/components/graphbuilder/GraphMessages.scala]: Done

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]

**Pull request merged**